### PR TITLE
SNOW-1458133 Implement `Index.value_counts`

### PR DESCRIPTION
**Pull request by sfc-gh-vbudati**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1458133

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Implemented `Index.value_counts` reusing the `Series.value_counts` implementation.

Examples:
```py
>>> import pandas as pd
>>> idx = pd.Index(
...         [1, 2, 3, 2, 3, 5, 6, 7, 8, 4, 4, 5, 6, 7, 1, 2, 1, 2, 3, 4, 3, 4, 5, 6, 7]
...     )
>>> idx.value_counts()
2    4
3    4
4    4
1    3
5    3
6    3
7    3
8    1
Name: count, dtype: int64
```

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 1]
```diff
@@ -10171,12 +10171,13 @@ def value_counts(
         ascending: bool = False,
         bins: Optional[int] = None,
         dropna: bool = True,
+        include_index: bool = False,
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10174]: @sfc-gh-vbudati since you start adding new interface for some of the query compiler interface now, how about let's do that for all APIs that requires extension to handle index, convert this function into a internal function like _value_counts(xxx). and you can leave the check for multiIndex to the index interface
```
 if include_index:
                if self.is_multiindex():
                    ErrorMessage.not_implemented(
                        "`value_counts` is not yet implemented for MultiIndex objects"
                    )
```

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 2]
```diff
@@ -1491,14 +1491,21 @@ def value_counts(
...
-        WarningMessage.index_to_pandas_warning("value_counts")
-        return self.to_pandas().value_counts(
-            normalize=normalize,
-            sort=sort,
-            ascending=ascending,
-            bins=bins,
-            dropna=dropna,
+        if bins is not None:
+            ErrorMessage.not_implemented(
+                "`bins` is not yet supported for Index.value_counts."
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:1496]: once you have a separate query_compiler interface for index, please move this check to the query compiler. Wether a parameter is currently supported or not is the responsibility of backend

File: tests/integ/modin/index/test_index_methods.py [hunk 3]
```diff
@@ -14,8 +14,71 @@
...
+    ),
+    native_pd.Index([1.1, 2.2, 1.0, 1, 1.1, 2.2, 1, 1, 1, 2, 2, 2, 2.2]),
+    native_pd.Index([1, 3, 1, 1, 1, 3, 1, 1, 1, 2, 2, 2, 3], name="random name"),
+    native_pd.Index(
+        [True, False, True, False, True, False, True, False, True, True], dtype=bool
+    ),
+    native_pd.Index(
+        [
+            "a",
+            "b",
```
Comment (sfc-gh-yzou) [inline tests/integ/modin/index/test_index_methods.py:54]: do you have similar test as what you have fixed? can you run this with jenkins? in fact, for when sort = True, and the result have a tie it is probably better to always jsut compare with the hand coded expected result

Comment (sfc-gh-vbudati) [reply tests/integ/modin/index/test_index_methods.py]: Yeah, I copied over the jenkins test here

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -10163,6 +10163,108 @@ def resample(
...
+                "`value_counts` is not yet implemented for MultiIndex objects"
+            )
+        by = self._modin_frame.index_column_pandas_labels
+
+        # validate whether by is valid (e.g., contains duplicates or non-existing labels)
+        self.validate_groupby(by=by, axis=0, level=None)
+
+        # append a dummy column for count aggregation
+        COUNT_LABEL = "value_count"
+        query_compiler = SnowflakeQueryCompiler(
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10211]: Here we seems duplicating the code in snowflake_query_compiler.value_counts, which is not a good practice, can you please extract the common code out to a common function used by both interface, and calls into it accordingly?

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 5]
```diff
@@ -1426,21 +1426,20 @@ def value_counts(
         normalize: bool = False,
         sort: bool = True,
         ascending: bool = False,
-        bins: Any = None,
+        bins: int | None = None,
         dropna: bool = True,
-    ) -> native_pd.Series:
-        # how to change the above return type to modin pandas series?
+    ) -> Series:
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:1431]: Is there a reason why we're not adding the docstrings under the docstrings directory to override Modin's docstrings similar to Series, DataFrame, ..etc?

File: CHANGELOG.md [hunk 6]
```diff
@@ -62,6 +62,11 @@
 - Added support for `DataFrame.corr`.
 - Added support for `DataFrame.equals` and `Series.equals`.
 - Added support for `DataFrame.reindex` and `Series.reindex`.
+- Added support for `Index.T`.
```
Comment (sfc-gh-vbudati) [inline CHANGELOG.md:65]: I added these here since I think the actual entries for these got wiped out by some other merge?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 7]
```diff
@@ -10659,6 +10659,46 @@ def resample(
...
+                Sort in ascending order.
+            bins : int, optional
+                Rather than count values, group them into half-open bins,
+                a convenience for ``pd.cut``, only works with numeric data.
+                This argument is not supported yet.
+            dropna : bool, default True
+                Don't include counts of NaN.
+        """
+        if bins is not None:
+            raise ErrorMessage.not_implemented("bins argument is not yet supported")
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10693]: Can you add a test to verify that we raise a NotImplementedError in this case?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 8]
```diff
@@ -10659,6 +10659,46 @@ def resample(
...
+
+        if self.is_multiindex():
+            ErrorMessage.not_implemented(
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10696]: Same as above.

File: docs/source/modin/supported/index_supported.rst [hunk 9]
```diff
@@ -125,7 +125,7 @@ Methods
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``nunique``                 | Y                               |                                  |                                                    |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
-| ``value_counts``            | N                               |                                  |                                                    |
+| ``value_counts``            | Y                               |                                  |                                                    |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/index_supported.rst:128]: This should be P because there are cases where we raise NotImplementedError. Please document them here.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 8 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10696]: This error can't be triggered right now since we don't have a MultiIndex object type implemented and the current Index object supports only one column. So I can't write a test for it. 😅

File: CHANGELOG.md (hunk 6 above)
Comment (sfc-gh-yzou) [reply CHANGELOG.md:76]: Oh, @sfc-gh-vbudati are those missing in the current release log? you will have to cherry-pick that into release-v1.20.0

File: CHANGELOG.md [hunk 10]
```diff
@@ -73,6 +73,11 @@
...
+- Added support for `Index.astype`.
+- Added support for `Index.unique` and `Index.nunique`.
+- Added support for `Index.value_counts`.
```
Comment (sfc-gh-yzou) [inline CHANGELOG.md:79]: new release have been made, so please add the new change into the new merges

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 8 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10712]: That just seems a dead code here, you probably just want to add an assert, assert not self.is_multiindex(), since you don't expect multiIndex object can call into this function

File: CHANGELOG.md (hunk 6 above)
Comment (sfc-gh-vbudati) [reply CHANGELOG.md:76]: yeah, they are missing in the current release log

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 11]
```diff
@@ -10719,6 +10759,37 @@ def value_counts(
...
+        self,
+        by: Union[List[Hashable], Tuple[Hashable, ...]],
+        normalize: bool,
+        sort: bool,
+        ascending: bool,
+        dropna: bool,
+    ) -> "SnowflakeQueryCompiler":
+        """
+        Helper method to obtain the frequency or number of unique values
+        a given SnowflakeQueryCompiler using groupby.
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10774]: -> "obtain the frequency of number of unique values within a group"

File: src/snowflake/snowpark/modin/plugin/extensions/index.py (hunk 5 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:1438]: Since this is a new class which doesn't exist in modin, I am not sure if the approach will work, but @sfc-gh-vbudati can you create a ticket to give it a try, it would be great if we can centralize all documentations

Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:1438]: https://snowflakecomputing.atlassian.net/browse/SNOW-1542483 ticket here, I will try working on this soon

File: CHANGELOG.md (hunk 10 above)
Comment (sfc-gh-vbudati) [reply CHANGELOG.md:79]: I only moved the value_counts log, I left the others in the 1.20 release since they were merged before the release.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:10774]: I think it's supposed to be "or", not "of" since we are only returning the counts of each value. Counts can also be called frequency. I can rewrite this if you prefer

File: CHANGELOG.md [hunk 12]
```diff
@@ -73,6 +74,10 @@
 - Added support for `DataFrame.corr`.
 - Added support for `DataFrame.equals` and `Series.equals`.
 - Added support for `DataFrame.reindex` and `Series.reindex`.
+- Added support for `Index.T`.
```
Comment (sfc-gh-yzou) [inline CHANGELOG.md:77]: actually, @sfc-gh-vbudati when is Index.T added? is it in an older version or 1.20.0

Comment (sfc-gh-yzou) [reply CHANGELOG.md:77]: I double checked, index.T seems have been added in 1.19.0, can you add it at the correct place?

Comment (sfc-gh-vbudati) [reply CHANGELOG.md]: added

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 13]
```diff
@@ -1433,21 +1433,20 @@ def value_counts(
         normalize: bool = False,
         sort: bool = True,
         ascending: bool = False,
-        bins: Any = None,
+        bins: int | None = None,
```
Comment (sfc-gh-evandenberg) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:1436]:
```suggestion
        bins: Optional[int] = None,
```

File: CHANGELOG.md [hunk 14]
```diff
@@ -77,6 +78,10 @@
 - Added support for `DataFrame.corr`.
 - Added support for `DataFrame.equals` and `Series.equals`.
 - Added support for `DataFrame.reindex` and `Series.reindex`.
+- Added support for `Index.astype`.
```
Comment (sfc-gh-evandenberg) [inline CHANGELOG.md:81]: Is this related to this pr?  I don't see any code changes related to 'Index.astype'

File: CHANGELOG.md [hunk 15]
```diff
@@ -151,7 +156,7 @@
 - Added support for `DataFrame.stack`.
 - Added support for `DataFrame.pivot` and `pd.pivot`.
 - Added support for `DataFrame.to_csv` and `Series.to_csv`.
-- Added support for `Index.sort_values`.
+- Added support for `Index.T`.
```
Comment (sfc-gh-evandenberg) [inline CHANGELOG.md:159]: Is this related to this pr?  I don't see any code changes related to this.

File: CHANGELOG.md (hunk 14 above)
Comment (sfc-gh-yzou) [reply CHANGELOG.md:81]: @sfc-gh-evandenberg it is not related, it is a fix of the previous wrong change log conflict resolving. @sfc-gh-vbudati can you actually put that in a different pr, we can merge it separately

File: CHANGELOG.md [hunk 16]
```diff
@@ -77,6 +78,10 @@
 - Added support for `DataFrame.corr`.
 - Added support for `DataFrame.equals` and `Series.equals`.
 - Added support for `DataFrame.reindex` and `Series.reindex`.
+- Added support for `Index.astype`.
+- Added support for `Index.unique` and `Index.nunique`.
+- Added support for `Index.sort_values`.
```
Comment (sfc-gh-yzou) [inline CHANGELOG.md:83]: that should be removed, it is for 1.20.0 and is listed above already

File: CHANGELOG.md (hunk 14 above)
Comment (sfc-gh-vbudati) [reply CHANGELOG.md]: I'll put this in a different PR, here: https://github.com/snowflakedb/snowpark-python/pull/1939

File: src/snowflake/snowpark/modin/plugin/extensions/index.py (hunk 13 above)
Comment (sfc-gh-vbudati) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:1436]: The linter changes it to `int | None` whenever I use `Optional`

**Review by sfc-gh-yzou — APPROVED** [review]

File: docs/source/modin/supported/index_supported.rst [hunk 17]
```diff
@@ -5,6 +5,8 @@ The following table is structured as follows: The first column contains the meth
 The second column is a flag for whether or not there is an implementation in Snowpark for
 the method in the left column.
 
+Currently, there is no MultiIndex support.
```
Comment (sfc-gh-joshi) [inline docs/source/modin/supported/index_supported.rst:8]: Does this mean we don't support any operations on MultiIndex objects? We definitely do have APIs that return them, at least.

**Review by sfc-gh-joshi — COMMENTED** [review]
one minor q

File: docs/source/modin/supported/index_supported.rst (hunk 17 above)
Comment (sfc-gh-vbudati) [reply docs/source/modin/supported/index_supported.rst:8]: Nope, we don't support any operations on MultiIndex since we don't have an object representing them. The lazy Index object we currently have is only supposed to be a single Index.

Comment (sfc-gh-joshi) [reply docs/source/modin/supported/index_supported.rst:8]: I think this is still worth clarifying, since it might imply to users that DF/Series operations cannot return MultiIndex objects, even though they're just returning eager pandas.MultiIndex.

Comment (sfc-gh-vbudati) [reply docs/source/modin/supported/index_supported.rst:8]: I suppose I'll say that there is no lazy MultiIndex support. This Index object is supposed to only be a single Index but existing DataFrame and Series APIs may support native pandas MultiIndex operations.

**Review by sfc-gh-joshi — APPROVED** [review]

File: docs/source/modin/supported/index_supported.rst (hunk 17 above)
Comment (sfc-gh-vbudati) [reply docs/source/modin/supported/index_supported.rst]: should be in now if you want to look at the text

**Review by sfc-gh-azhan — APPROVED** [review]
